### ISSUE TITLE
amp-story: Images with missing or inadequate text descriptions

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#32493]
# Images with missing or inadequate text descriptions

## WCAG Level
Level A

## Priority
High

## Pages/screens/components affected
* [Navigation](https://www.gstatic.com/amphtml/stamp/qa/navigation.html)
* [RTL](https://www.gstatic.com/amphtml/stamp/qa/rtl.html)
* [Video](https://www.gstatic.com/amphtml/stamp/qa/video.html)
* [Audio - amp-story background audio](https://www.gstatic.com/amphtml/stamp/qa/audio-story.html)
* [Audio - amp-story-page background audio](https://www.gstatic.com/amphtml/stamp/qa/audio-page.html)
* [Audio - amp-story using amp-audio](https://www.gstatic.com/amphtml/stamp/qa/amp-audio.html)
* [Animation](https://www.gstatic.com/amphtml/stamp/qa/animation.html)
* [Sharing](https://www.gstatic.com/amphtml/stamp/qa/sharing.html)
* [Tooltip](https://www.gstatic.com/amphtml/stamp/qa/tooltip.html)
* [Page attachment](https://www.gstatic.com/amphtml/stamp/qa/page-attachment.html)
* [Branching](https://www.gstatic.com/amphtml/stamp/qa/branching.html#page=page2)
* [Consent](https://www.gstatic.com/amphtml/stamp/qa/consent.html)
* [Prerender - Image](https://www.gstatic.com/amphtml/stamp/qa/prerender-img.html?amp_js_v=0.1#origin=https://www.google.com&prerenderSize=1&visibilityState=prerender&paddingTop=0&p2r=0&horizontalScrolling=0&csi=1&aoh=15526927198732&viewerUrl=https://www.google.com&history=1&storage=1&cid=1&cap=swipe,navigateTo,cid,fragment,replaceUrl)
* [Prerender - Video](https://www.gstatic.com/amphtml/stamp/qa/prerender-video.html?amp_js_v=0.1#origin=https://www.google.com&prerenderSize=1&visibilityState=prerender&paddingTop=0&p2r=0&horizontalScrolling=0&csi=1&aoh=15526927198732&viewerUrl=https://www.google.com&history=1&storage=1&cid=1&cap=swipe,navigateTo,cid,fragment,replaceUrl)
* [AMP Story Player - Player](https://www.gstatic.com/amphtml/stamp/qa/player.html)
* [AMP Story Player - AMP Story 360](https://www.gstatic.com/amphtml/stamp/qa/amp-story-360.html)

**List of elements that need text descriptions added/revised:**
- [x] Pagination buttons desktop #32861
- [x] Share menu close button #32863
- [x] Embedded tweet close button #32863
- [x] amp-story-360  #32871
- [x] amp-video #32963
- [x] amp-img #32880
- [x] Page attachment close button #32925

## Description
When images do not have a text description it means anyone who cannot see images properly will not know what the images are for.

Across the tested samples, we found recurring image-based elements (such as the previous/next/replay navigation controls, and the "X" close control for the sharing modal dialog), as well as individual content images in the stories themselves, which lack an appropriate text alternative.

## User impact
If an image does not have an appropriate text description, someone who is blind and uses a screen reader or someone who has images turned off in their browser to save on bandwidth will be completely unaware of the image's purpose. If the image conveys important information like a graph, or is critical to the operation of a control like a link or button, then a missing text description can prevent them from completing vital tasks.

## Required solution
Make sure that every image that conveys information or is used to perform an action has an appropriate text description. The text description must be clear and relevant to the context in which the image is being used.

This solution must be applied to all instances of the issue identified within the test sample, then applied to all other instances of the same issue identified throughout the rest of the website.

### Graphical controls lacking a text alternative

Across all tested samples, the graphical previous/next/replay controls visible in the large-screen/desktop version of AMP Stories are implemented as `<div>` elements wrapped around a `<button>`.

![One of the 'next' buttons](https://user-images.githubusercontent.com/54933393/107243144-9ff60700-6a24-11eb-8094-80cfc91429c1.png)

```
<div class="i-amphtml-story-button-container i-amphtml-story-fwd-next next-container" role="button" aria-label="Next page">
  <button class="i-amphtml-story-button-move"></button>
</div>
```

An attempt has been made to provide a text alternative for these buttons by providing an `aria-label` on the wrapping `<div>`. However, this is not consistently announced by assistive technologies - particularly when navigating directly to controls using `Tab`/`Shift`+`Tab` (which sets focus directly to the child `<button>` element, which lacks an accessible name).

The current approach was likely taken because in small-screen/smartphone view, the actual `<button>` is removed. However, we do not recommend this approach (see the related best practice issue to add actual focusable and visible navigation controls even in small-screen view). We recommend to instead move the `aria-label` to the `<button>`, or to at least duplicate it (having it both on the `<div>` and the `<button>`).

```
<div ...>
  <button class="i-amphtml-story-button-move" aria-label="Next page"></button>
</div>
```

In the sharing modal dialog (in large-screen/desktop view), the "X" close control (which is currently not keyboard-accessible - see the separate issue on keyboard access) is currently announced as "times" by assistive technologies.

![The sharing modal dialog, with its visible 'x' close control](https://user-images.githubusercontent.com/54933393/107243231-b43a0400-6a24-11eb-9bba-526623d8b749.png)

```
<span class="i-amphtml-story-share-menu-close-button" role="button">×</span>
```

Similarly, in [Tooltip](https://www.gstatic.com/amphtml/stamp/qa/tooltip.html), when expanding the embedded tweet, the "X" close control lacks a text alternative (and is also currently not keyboard-operable).

![The 'X' close control when the embedded tweet is expanded](https://user-images.githubusercontent.com/54933393/107243288-c4ea7a00-6a24-11eb-9b66-1f9ccf9a1024.png)

```
<span class="i-amphtml-expanded-view-close-button" role="button"></span>
```

Beyond changing these control to an actual `<button>`, we recommend providing an `aria-label` that explicitly conveys the meaning/purpose of the controls.

```
<button class="i-amphtml-story-share-menu-close-button" aria-label="Close">×</button>
```

```
<button class="i-amphtml-expanded-view-close-button" aria-label="Close"></button>
```

### Graphical content lacking a text alternative

The sample AMP stories contain various instances of images, video elements and canvas-based animations that don't provide a text alternative. It is worth noting that while these images, videos and animations are only used for demonstration purposes in these pages, they should be given appropriate text alternatives, the same way that real-world content in an actual AMP story would (unless they are used purely for decorative purposes).

[Navigation](https://www.gstatic.com/amphtml/stamp/qa/navigation.html) contains good examples of video elements which currently lack any kind of text alternative (as well as not providing any audio, nor any transcript - see the separate issue *Missing or inaccurate transcript*).

![An AMP Story page with a video element](https://user-images.githubusercontent.com/54933393/107243356-d469c300-6a24-11eb-9b67-70cfbb1f5894.png)

```
<amp-video ... alt="Person running over a bridge" ...>
  <video ...>...</video>
  ...
</amp-video>
```

While the `<amp-video>` element has been given an `alt="..."` attribute, this is not conveyed to assistive technologies (as it is non-standard on a custom element). In this case, consider generating a visually-hidden `<div>` or similar element for the `<amp-video>` that contains the text alternative.

```
<amp-video ...>
  <video ...>...</video>
  ...
  <div class="visually-hidden">Person running over a bridge</div>
</amp-video>
```

```
.visually-hidden {
  clip: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}
```

Another approach is to add the alternative text as an `aria-label` on the `<video>` element.

```
<amp-video ...>
  <video ... aria-label="Person running over a bridge">...</video>
  ...
</amp-video>
```

[AMP Story Player - AMP Story 360](https://www.gstatic.com/amphtml/stamp/qa/amp-story-360.html) features `<canvas>`-based 3D animations. These don't currently have any text alternative, meaning that their presence is not conveyed at all to assistive technology users.

![An AMP story with a 3D canvas animation](https://user-images.githubusercontent.com/54933393/107243436-e77c9300-6a24-11eb-9e41-b05e505841c8.png)

```
<amp-img src="assets/hello_world_360.jpg" ... alt=" ">
  <img ... alt=" " ...>
</amp-img>
<div class="i-amphtml-fill-content i-amphtml-replaced-content">
  <canvas width="410" height="684"></canvas>
</div>
```

As these `<canvas>`-based elements do convey content/meaning (even though they are only test/sample items in these cases), they should be provided with an appropriate text alternative. For the `<canvas>` element, this can be done using `aria-label`, and additionally giving the element an explicit `role="img"` to denote it as an image.

```
<canvas ... role="img" aria-label="Demonstration of a 3D animation: view moves from 'start here' to 'end here'">
```

[Animation](https://www.gstatic.com/amphtml/stamp/qa/animation.html) contains various examples of animations, including a "Ken Burns" style zoom-and-pan animation over an image.

![AMP story page with an image in the background](https://user-images.githubusercontent.com/54933393/107243482-f400eb80-6a24-11eb-9acc-bd82dfe10c1b.png)

```
<amp-img id="ken-burns-img1" src="assets/running.png" ... alt=" ">
  <img ... alt=" " ...>
</amp-img>
```

The image in this case has been given an empty `alt=" "` (though this should be completely empty, i.e. `alt=""`). In cases where these sorts of images/animations are purely decorative, this approach is acceptable. However, as in this case the image is effectively the focus of the actual page (as the page demonstrates the use of animations), it would be more appropriate to provide an actual text alternative that describes the purpose of the image, such as "Ken Burns animation over a still image".

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied to issues identified within the test sample, and to test the rest of the website for instances of the same issue:
1. Check that all images, videos, and meaningful visual elements/animations, have appropriate alternative text.
2. Check that the alternative text acts as a suitable equivalent for the image.

### Definition of done
Complete all of these tasks before closing this issue or indicating it is ready for retest:
* All issues identified within the test sample have been resolved.
* The rest of the website has been tested for the same issue.
* All issues identified throughout the rest of the website have been resolved or filed as new issues.

## Related standards
* [WCAG 2.1 Success Criterion 1.1.1 Non-text Content (Level A)](https://www.w3.org/TR/WCAG21/#non-text-content)

## More information
* [W3C: Image concepts](https://www.w3.org/WAI/tutorials/images/decision-tree/)
* [W3C: An alt decision tree](https://www.w3.org/WAI/tutorials/images/decision-tree/)
* [W3C: HTML 5.2 Requirements for providing text to act as an alternative for images](https://www.w3.org/TR/html52/semantics-embedded-content.html#alt-text)
* [The A11Y Project - How to: hide content](https://www.a11yproject.com/posts/2013-01-11-how-to-hide-content/)

## Test data
Test date: January 2021

**Comment by gmajoulet** [comment]
Actionable eng items:

1. Switch the `aria-label` to the `button` elements and not the wrapping `<div>`
2. Switch the elements that are not buttons yet to be buttons
3. Some of the content is publisher provided (eg: amp-video or amp-img), and we add a default `aria-label=" "` when none is provided, we should make it completely empty `aria-label=""`

**Comment by processprocess** [comment]
Closing this. Fixed via:

- Pagination buttons desktop #32861
- Share menu close button #32863
- Embedded tweet close button #32863
- amp-story-360 #32871
- amp-video #32963
- amp-img #32880
- Page attachment close button #32925
